### PR TITLE
Report bug2

### PR DESCRIPTION
**Pull request by tanguyduval**
New qMR_reportBug:
send bugs via emails
auto catch in Guis

### CODE REVIEW

**Comment by tanguyduval** [comment]
@mathieuboudreau use this PR ;)

**Comment by mathieuboudreau** [comment]
@tanguyduval I tried it, and am having trouble getting the email to send. I modified the code with my own email:

```

        setpref('Internet','E_mail','(MYEMAIL)@gmail.com')
        setpref('Internet','SMTP_Server','mail')
        txt = {}; for ii=1:size(txt2send,1), txt{ii} = txt2send(ii,:); end
        sendmail('(MYEMAIL)@gmail.com','qMRLab issue',sprintf('%s\n',txt{:}))

```

But am getting the following error:

```

Error using sendmail (line 169)
Unknown SMTP host: mail;
mail

Error in qMR_reportBug (line 36)
        sendmail('(MYEMAIL)@gmail.com','qMRLab issue',sprintf('%s\n',txt{:}))

Error in tryfunc (line 5)
    qMR_reportBug(err)

Error in addtryfunc>@(a,b)tryfunc(CB,a,b) (line 14)
                set(listobj(iuic),'Callback', @(a,b) tryfunc(CB, a, b))
 
Error while evaluating UIControl Callback.

```

Can you confirm that you tested it and got an email to send on your system? If so, I'll dig a bit deeper. From the forums I'm reading, they suggest setting the username and password for the "sender" in the MATLAB preferences, but this sounds like it's not an option for us.

**Comment by mathieuboudreau** [comment]
The current qMRLab commit number, OS, and MATLAB version would also be great information to report.

**Comment by jcohenadad** [comment]
i tried configuring email-based bug reports in the past but it become a nightmare, which is why we moved to Sentry. The issues that @mathieuboudreau has experienced, users (who are likely less experienced than Mathieu) will also experience them if we rely on their SMTP. So i am not sure this is the right way to go

**Comment by tanguyduval** [comment]
Hello hello, don't change the setpref. For @compagny.com you can use smtp serveur "mail" with no password ;)

**Comment by tanguyduval** [comment]
Agreed to add os and Matlab version

**Comment by tanguyduval** [comment]
@mathieuboudreau je ne vais pas avoir le temps de travailler la dessus les 2 prochaines semaines. Si tu veux prendre le lead sur cette branche pdt ce temps n'hésite pas ;)

**Comment by mathieuboudreau** [comment]
>Hello hello, don't change the setpref. For @compagny.com you can use smtp serveur "mail" with no password ;)

@tanguyduval I tried this first (by only changing), and it did not work for me on my OS (OSX 10.12.16) and MATLAB version (2017a).:

git diff:

<details>

```

mathieuboudreau@brassens:~/neuropoly/github/qMRLab$ git diff
diff --git a/src/Common/qMR_reportBug.m b/src/Common/qMR_reportBug.m
index 62cbfba..dd541a5 100644
--- a/src/Common/qMR_reportBug.m
+++ b/src/Common/qMR_reportBug.m
@@ -33,7 +33,7 @@ if any(~cellfun(@isempty,qmrexcep))
         setpref('Internet','E_mail','qMRLabbugreport@company.com')
         setpref('Internet','SMTP_Server','mail')
         txt = {}; for ii=1:size(txt2send,1), txt{ii} = txt2send(ii,:); end
-        sendmail('qmrlab_developers@googlegroups.com','qMRLab issue',sprintf('%s\n',txt{:}))
+        sendmail('emb6150@gmail.com','qMRLab issue',sprintf('%s\n',txt{:}))
     else
         disp('not reporting error...')
     end
mathieuboudreau@brassens:~/neuropoly/github/qMRLab$ 

```

</details><p>

Matlab error after submitting a bug report:

<details>

```

Warning: qMRLab error detected 
> In qMR_reportBug (line 13)
  In tryfunc (line 5)
  In addtryfunc>@(a,b)tryfunc(CB,a,b) (line 14) 
Error using sendmail (line 169)
Unknown SMTP host: mail;
mail

Error in qMR_reportBug (line 36)
        sendmail('emb6150@gmail.com','qMRLab issue',sprintf('%s\n',txt{:}))

Error in tryfunc (line 5)
    qMR_reportBug(err)

Error in addtryfunc>@(a,b)tryfunc(CB,a,b) (line 14)
                set(listobj(iuic),'Callback', @(a,b) tryfunc(CB, a, b))
 
Error while evaluating UIControl Callback.

```

>> 

</details><p>

If this works for you, maybe you have preferences already set up that you aren't aware of? Otherwise, maybe this trick only works on Windows?

**Comment by mathieuboudreau** [comment]
> @mathieuboudreau je ne vais pas avoir le temps de travailler la dessus les 2 prochaines semaines. Si tu veux prendre le lead sur cette branche pdt ce temps n'hésite pas ;)

Ok – I don't think there's any rush. If I find the time I might poke at it some more, let me know if you want to video call to chat about it when your schedule clears up.

**Comment by tanguyduval** [comment]
@mathieuboudreau that should work better now. Now using an smtp2go account.
We are limited to 16 mail per hour, which is a great limitation in fact ;)
In order to add bug report to an existing GUI.
Add the following lines at the end of the opening function:
````
% list all uicontrols
listobj = findobj(hObject,'-depth',100,'Type','UIControl');
% add try/catch to the callback
addtryfunc(listobj);
````

Best,
Tanguy

**Comment by tanguyduval** [comment]
@mathieuboudreau I added a few lines to avoid dealing twice with the same error.
If an exception has already been caught by qMRLab, it just skips the bug report.... until the user restarts matlab or clear the variables (I use a persistent variable for known bugs).

**Comment by mathieuboudreau** [comment]
I tried causing a bug, sent a report, but no email was sent/received to the qmrlab the google group email. Where to do we go to check the bug reports?

**Comment by tanguyduval** [comment]
@mathieuboudreau I haven't tested sending to qmrlab_developers@googlegroups.com
However I can see that your bugs report were succesfully sent and did not bounce...:
image

The admin of the google group should set the Post Permission to Public... or add qMRLabbugreport@company.com as a member. Who is the admin??

**Comment by mathieuboudreau** [comment]
@tanguyduval Ah ok! I didn't know about that dashboard. I don't know if it's a good idea to spam the google groups with bug reports – it would be better to grant us access to that dashboard if you can. That's how we also do it with Sentry.

**Comment by tanguyduval** [comment]
@mathieuboudreau shall we create an email account for qMRLab then?

**Comment by mathieuboudreau** [comment]
@tanguyduval That might be a good approach to test out. I tried logging in to the other dashboard, and although I could see that the emails were sent/received, I never figured out how to open them,

**Comment by mathieuboudreau** [comment]
@tanguyduval Ah ha, sorry about that. Sorry about the delay, I had my hands busy this Fall with the series of interactive notebooks, so not much extra time to doing core qMRLab dev. I will revisit this soon.

**Comment by agahkarakuzu** [comment]
@mathieuboudreau @tanguyduval do you feel strongly about merging this feature?

**Comment by tanguyduval** [comment]
@agahkarakuzu looks like you are not a big fan of sending errors by email??
I propose to report errors directly to GitHub using the `web` function. in my last commit, Issues are prefilled with 
- the error stack
- the error message
- Matlab version
- qMRLab version

Best

**Comment by mathieuboudreau** [comment]
My main worry about this branch, is annoyance by repetition. We need some kind of logger that if a specific error has been encountered before, the 1) the user doesn't see the error pop up again, and 2) the error isn't sent a second time. Because when I first tried testing this branch (albeit a long time ago), my workflow was interrupted a several times in a short duration with error pop ups, and I'd much rather (as a user) be completely unaware of them. That's why I like the [Sentry]() style of error reporting; once the user initially accepts to error track, they never get interrupted or bothered by it, only a message in the terminal is given to say when an an error has been reported.

I'm afraid that if we automate errors to be sent as issues, we will get flooded with them (partly because not everything is perfect with the GUI, and secondly because of repetition in terms of how many times a user might encounter the same error, and how many users in total will encounter them).

**Comment by tanguyduval** [comment]
@mathieuboudreau I added a logger a while ago to prevent repeated popups. Before, the log was stored in a persisent variable (lost in new Matlab session). I now 28a9083 store it in a Matlab pref using `setpref`.

**Comment by tanguyduval** [comment]
Using the current version, users need to create a GitHub account and complete the GitHub issue page.

**Comment by mathieuboudreau** [comment]
Great! @agahkarakuzu do you want to test out this branch (not a full review), i.e. do lots of stuff that will cause errors, and see how the branch feels in its current state (in terms of user experience)?

**Comment by agahkarakuzu** [comment]
Let's put this to vote in the next meeting.

**Comment by tanguyduval** [comment]
Might be interesting to look at what the conn toolbox is doing for bug
report.

**Comment by agahkarakuzu** [comment]
I gave this one a try following Tanguy's example and I was able to reproduce it: 
image

GitHub issue is a much better (and less noisy) approach then auto-emails. If this is the only intended output, then 💚 ✅.

File: src/Common/qMR_reportBug.m [hunk 1]
```diff
@@ -0,0 +1,64 @@
+function qMR_reportBug(exception)
+% report bug to the qMRLab dev team.
```
Comment (jcohenadad) [inline src/Common/qMR_reportBug.m]: i would specify in the docstring that this function opens an issue on Github website

File: src/Common/qMR_reportBug.m [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                     'Error message:\n',...
+                     '`%s`\n',...
+                     '\nStack:\n',...
+                     '%s\n'],...
+                     qMRLabVer,version,message,stack);
+    body = urlencode(body);
+    title = '[USER]';
+    
+    % report
+    answer = questdlg('Report Bug on GitHub?','OUPS... A BUG OCCURED','Yes','No','Yes');
```
Comment (jcohenadad) [inline src/Common/qMR_reportBug.m]: in English: "OOPS"

Comment (agahkarakuzu) [reply src/Common/qMR_reportBug.m]: :)

Comment (jcohenadad) [reply src/Common/qMR_reportBug.m]: OCCURED --> OCCURRED

**Pull request merged**